smooth_l1_loss

sushmanthreddy opened the pull request:
close #21165

AryanSharma21 commented:
You can read about the tests in the documentations also 

https://unify.ai/docs/ivy/docs/ivy_tests.test_ivy.helpers.html

sushmanthreddy commented:
@zaeemansari70 can u plss review this request and merge it?

zaeemansari70 reviewed (changes requested):
Hi @sushmanthreddy , I've fixed the small issue locally which I mentioned below to test your code. The failures are as follows.

```
FAILED ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py::test_smooth_l1_loss[cpu-numpy-False-False] - KeyError: 'cpu'
FAILED ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py::test_smooth_l1_loss[cpu-jax-False-False] - ivy.utils.exceptions.IvyBackendException: jax: smooth_l1_loss: jax: subtract: sub does not accept dtype bool at position 0. Accepted dtypes at position 0 are subtypes of complex...
FAILED ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py::test_smooth_l1_loss[cpu-torch-False-False] - ivy.utils.exceptions.IvyBackendException: torch: smooth_l1_loss: torch: subtract: Subtraction, the `-` operator, with two bool tensors is not supported. Use the `^` or `logical_...
FAILED ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py::test_smooth_l1_loss[cpu-paddle-False-False] - AssertionError
```
All the backends seems to be failing, kindly work on a fix for these after fixing the issue which I mentioned below, Thanks! :slightly_smiling_face:

sushmanthreddy commented:
@zaeemansari70 can u pls review it once now??

sushmanthreddy commented:
@AryanSharma21 @zaeemansari70 can u plss review this once?? here is the screen shoot of passing test cases

![Screenshot from 2023-08-16 22-47-30](https://github.com/unifyai/ivy/assets/73489688/8991e4e8-1e37-469c-849a-55964ca1f362)


it is passing test cases in my local environment..can u pls review the pull request and give some feedback?

zaeemansari70 reviewed (changes requested):
Hey @sushmanthreddy ,
Your functional implementation seems to be failing for the falsifying example which I posted below:
```
E           ivy.utils.exceptions.IvyBackendException: paddle: smooth_l1_loss: paddle: where: (NotFound) The kernel with key (CPU, Undefined(AnyLayout), bfloat16) of kernel `where` is not registered. Selected wrong DataType `bfloat16`. Paddle support following DataTypes: int32, int64, float64, float32.
E             [Hint: Expected kernel_iter == iter->second.end() && kernel_key.backend() == Backend::CPU != true, but received kernel_iter == iter->second.end() && kernel_key.backend() == Backend::CPU:1 == true:1.] (at /paddle/paddle/phi/core/kernel_factory.cc:199)
E           
E           Falsifying example: test_smooth_l1_loss(
E               backend_fw='paddle',
E               on_device='cpu',
E               dtype_and_true=(['bfloat16'], [array([-1, -1, -1], dtype=bfloat16)]),
E               dtype_and_pred=(['bfloat16'], [array([-1, -1, -1], dtype=bfloat16)]),
E               beta=0.5,
E               reduction='none',
E               test_flags=FunctionTestFlags(
E                   ground_truth_backend='tensorflow',
E                   num_positional_args=2,
E                   with_out=False,
E                   instance_method=False,
E                   test_gradients=None,
E                   test_compile=None,
E                   as_variable=[False],
E                   native_arrays=[False],
E                   container=[False],
E               ),
E               fn_name='smooth_l1_loss',
E           )
E           
E           You can reproduce this example by temporarily adding @reproduce_failure('6.82.4', b'AXicY2BkAAMoBaZxsYlTdIABHQAAEPkAzg==') as a decorator on your test case
```

sushmanthreddy commented:
@zaeemansari70 The stateful tests are failing because of a problem with the testing pipeline for the stateful API which as far as I know hasn't been fixed yet. The paddle test is failing because of an error with the bfloat16 datatype which again is a known issue that Ivy people are working on. 
![Screenshot from 2023-08-17 00-36-58](https://github.com/unifyai/ivy/assets/73489688/ca6c5b06-b785-442e-982c-555d89f16b3a)

see these are result of example loss functionswhich is failing for paddle,so pls review it once again

sushmanthreddy commented:
@zaeemansari70 this issue is about adding backend and ivy  experimental functional api .So i have added the api  in ivy experimental folder and added the respectiive backend of it.And smooth_l1_loss is only supported in torch  and paddle .so I have added the backends and it is smooth_l1_loss is a mixed function .

And failure of tests, The stateful tests are failing because of a problem with the testing pipeline for the stateful API which as far as I know hasn't been fixed yet. The paddle test is failing because of an error with the bfloat16 datatype which again is a known issue that Ivy people are working on.
 
i hope this can be reviwed once again.I have provided the test cases screen shot above.so pls provide some feed backover it

zaeemansari70 commented on ivy/functional/backends/paddle/experimental/losses.py:
```diff
@@ -0,0 +1,52 @@
+
+
+@with_unsupported_device_and_dtypes(
+    {"2.5.1 and below": {"cpu": ("float16",)}}, backend_version
```
The `unsupported_dtypes` will be added here in the decorator.
They should not be specified separately.
Kindly remove the list you added above.

zaeemansari70 reviewed (changes requested):
Requested some changes below. :+1: 

Also kindly uncommit all your `stateful` changes, the issue you linked to is a extension of the ivy API todolist.

Thanks!

zaeemansari70 commented on ivy/functional/ivy/experimental/losses.py:
```diff
@@ -23,7 +24,7 @@ def log_poisson_loss(
     *,
     compute_full_loss: bool = False,
     axis: int = -1,
-    reduction: str = "none",
```
kindly uncommit this change too

sushmanthreddy replied:
this commit is about log_poission_loss..should i really change,these changes are made by someone..

zaeemansari70 replied:
The PR shows that you have updated the value of `reduction` to "mean", the only changes that should be shown should be just for `smooth_l1_loss` kindly un-commit the change which is being showed :slightly_smiling_face:

zaeemansari70 commented on ivy/functional/backends/torch/experimental/losses.py:
```diff
@@ -0,0 +1,40 @@
+import torch
+import torch.nn.functional as F
```
You should just import `torch`

zaeemansari70 commented on ivy/functional/backends/torch/experimental/losses.py:
```diff
@@ -0,0 +1,40 @@
+    reduce: Optional[bool] = None,
+    out: Optional[torch.Tensor] = None,
+) -> torch.Tensor:
+    return F.smooth_l1_loss(
```
and use it like this `torch.nn.functional.smooth_l1_loss()`

zaeemansari70 commented on ivy/functional/backends/paddle/experimental/losses.py:
```diff
@@ -0,0 +1,41 @@
+# global
+from typing import Optional
+import paddle
+import paddle.nn.functional as F
```
same instructions for paddle as given below for torch

zaeemansari70 replied:
You haven't addressed this yet, you have to use paddle the same way you used torch

zaeemansari70 reviewed (changes requested):
I've requested some changes 
Also please uncommit your `libssl` file which I assume you have mistakenly added.

Thanks!

sushmanthreddy commented:
![Screenshot from 2023-08-20 02-27-47](https://github.com/unifyai/ivy/assets/73489688/047accd9-4249-4384-b722-c79793f40dea) 
checked supporting dtypes by this method.

zaeemansari70 reviewed (changes requested):
2 small changes which you missed to address before

sushmanthreddy commented:
@zaeemansari70 i havee closed this  becuase of so many merge conflicts
i have made pull request ,if posssible plss assign to yourself and review it..they are basically same code from the last pr.
https://github.com/unifyai/ivy/pull/22378
